app/services/adviser/matchback.rb

```diff
@@ -0,0 +1,27 @@
+module Adviser::Matchback
+  extend ActiveSupport::Concern
+
+private
+
+  def matchback_candidate
+    @matchback_candidate ||= begin
+      api = GetIntoTeachingApiClient::TeacherTrainingAdviserApi.new
+      api.matchback_candidate(existing_candidate_request)
+    rescue GetIntoTeachingApiClient::ApiError => e
+      #  A 404 not found is returned when the matchback is unsuccessful,
+      # indicating that the candidate does not exist in the GiT API.
+      raise unless e.code == 404
+
+      nil
+    end
+  end
+
+  def existing_candidate_request
+    GetIntoTeachingApiClient::ExistingCandidateRequest.new({
+      email: application_form.candidate.email_address,
+      first_name: application_form.first_name,
+      last_name: application_form.last_name,
+      date_of_birth: application_form.date_of_birth,
+    })
+  end
+end
```

Comment:
I'd move this module to a class. Something like `MatchbackCandidate.new(application_form:).call` which call the relevant things in the api and returns an object with attributes transformed. Or in case there is an error it returns an rich object to do verification if there is any error on the API. Then you can use on the SignupAvailability and in the worker too and remove the shared example altogether. This will make more simple and easier to read. @carlosmartinez wdyt on this case?

Reply:
I've changed the module to be a class and swapped the transformer class for a decorator that can be returned from the candidate matchback method. I'm not quite sure I follow the changes to the error response but happy to jump on a call if you want to talk me through it?